### ISSUE TITLE
question: How to setup EventBridge in localstack?

### ISSUE THREAD

**Issue by Vuta** [localstack/localstack#6156]
### Is there an existing issue for this?

- [X] I have searched the existing issues and read the documentation

### Question

Hi everyone, I have a question regarding the EventBridge. I tried with both docker and manually build from source locally, they both gave me error that I will describe below.

First I created a connection
```bash
awslocal events create-connection \
  --name "API_Connection_1" \
  --description "Local endpoint that receives the event" \
  --authorization-type "API_KEY" \
  --auth-parameters '{ "ApiKeyAuthParameters": { "ApiKeyName": "X-API-KEY", "ApiKeyValue": "api_key" } }'

# the response
{
    "ConnectionArn": "arn:aws:events:us-east-1:000000000000:connection/API_Connection_1/36285233-0b81-45c8-a28f-45fa6e43a395",
    "ConnectionState": "AUTHORIZED",
    "CreationTime": 1653619840.268913,
    "LastModifiedTime": 1653619840.268913
}
```

Next, I created an API destination (which points to my development web server `http://localhost:3300/test`)
**Important: My endpoint just returns a `hello` string**
```bash
awslocal events create-api-destination \
  --name "destination_1" \
  --description "send test event to local server" \
  --connection-arn "arn:aws:events:us-east-1:000000000000:connection/API_Connection_1/36285233-0b81-45c8-a28f-45fa6e43a395" \
  --invocation-rate-limit-per-second 50  \
  --invocation-endpoint "http://localhost:3300/test" \
  --http-method "GET"

# the response
{
    "ApiDestinationArn": "arn:aws:events:us-east-1:000000000000:api-destination/destination_1/c05ee5b9-4a7a-4044-8652-a42950b273a3",
    "ApiDestinationState": "ACTIVE",
    "CreationTime": 1653620018.345984,
    "LastModifiedTime": 1653620018.345984
}
```
However, checking the localstack logs gave me this
```bash
# I removed all un-related messages
IN(events): "POST /"  - data: b'{
  "Name": "destination_1", 
  "Description": "send test event to local server", 
  "ConnectionArn": "arn:aws:events:us-east-1:000000000000:connection/API_Connection_1/1d148dc5-1030-4cb4-b618-8e3b5c4d7d60", 
  "InvocationEndpoint": "hello"**, 
  "HttpMethod": "GET", 
  "InvocationRateLimitPerSecond": 50}'
```

As you can see, the response's `InvocationEndpoint` is `hello`, which is what my API endpoint returns.
This caused a problem because later on, when I tried to put an event, localstack used `hello` as the destination and gave me this error
```bash
Calling EventBridge API destination (state "ACTIVE"): GET hello
Unable to send event notification 
{'version': '0', 'id': 'eab8e770-bec1-467f-9733-c3373502c8d5', 'detail-type': 'test', 'source': 'loc... to target 
{'Id': 'Test_Destination_1', 'Arn': 'arn:aws:events:us-east-1:000000000000:api-destination/destination_1/e6b7ea19-319f-4273-86a6-dbaa6a1baa87'}: 
Invalid URL 'hello': No schema supplied. Perhaps you meant http://hello?
```

To confirm my suspicion, I updated my local endpoint to return it's own URL, and indeed I was able to put and receive the event after that.

So now my questions are:
1. Am I missing any super obvious steps to setup here?
2. If this is the expected behavior of localstack (which I doubt of), I wonder why though? 

Would really appreciate any help on this, thank in advance 🙏

**Comment by corinz** [comment]
+1 and 6 thumbs up on this question. I am experiencing the same issue. Can we get a triage? @silv-io

**Comment by gage-russell** [comment]
+1

**Comment by corinz** [comment]
@dominikschubert Can this be triaged?

**Comment by gustavopedro20** [comment]
Same error here.

**Comment by bentsku** [comment]
Hello!

I've tried to reproduce the issue, and it seems this is an `awscli` issue. If you pass a link as a parameter, it will try to fetch it (common usage with `file://` and such).

TLDR; the fix is to enclose the address of the endpoint in single quotes as well as double quotes:

```bash
$ awslocal --debug events create-api-destination \
  --name "destination_1" \
  --description "send test event to local server" \
  --connection-arn "arn:aws:events:us-east-1:000000000000:connection/API_Connection_1/36285233-0b81-45c8-a28f-45fa6e43a395" \
  --invocation-rate-limit-per-second 50  \
  --invocation-endpoint "'http://localhost:3300/test'" \
  --http-method "PUT"

```

### More explanations

While passing the parameters like you explained, I had another issue because I didn't have a web server running locally.

I tried running the following command with the `--debug` flag:
```bash
$ awslocal --debug events create-api-destination \
  --name "destination_1" \
  --description "send test event to local server" \
  --connection-arn "arn:aws:events:us-east-1:000000000000:connection/API_Connection_1/36285233-0b81-45c8-a28f-45fa6e43a395" \
  --invocation-rate-limit-per-second 50  \
  --invocation-endpoint "http://localhost:3300/test" \
  --http-method "PUT"

```
The result was:

```
[...cutting as the traceback is very long]
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/awscli/clidriver.py", line 218, in main
    return command_table[parsed_args.command](remaining, parsed_args)
  File "/usr/local/lib/python3.9/site-packages/awscli/clidriver.py", line 362, in __call__
    return command_table[parsed_args.operation](remaining, parsed_globals)
  File "/usr/local/lib/python3.9/site-packages/awscli/clidriver.py", line 504, in __call__
    call_parameters = self._build_call_parameters(
  File "/usr/local/lib/python3.9/site-packages/awscli/clidriver.py", line 560, in _build_call_parameters
    value = self._unpack_arg(arg_object, value)
  File "/usr/local/lib/python3.9/site-packages/awscli/clidriver.py", line 571, in _unpack_arg
    return unpack_argument(session, service_name, operation_name,
  File "/usr/local/lib/python3.9/site-packages/awscli/argprocess.py", line 79, in unpack_argument
    value_override = session.emit_first_non_none_response(
  File "/usr/local/lib/python3.9/site-packages/botocore/session.py", line 764, in emit_first_non_none_response
    responses = self._events.emit(event_name, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/botocore/hooks.py", line 412, in emit
    return self._emitter.emit(aliased_event_name, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/botocore/hooks.py", line 256, in emit
    return self._emit(event_name, kwargs)
  File "/usr/local/lib/python3.9/site-packages/botocore/hooks.py", line 239, in _emit
    response = handler(**kwargs)
  File "/usr/local/lib/python3.9/site-packages/awscli/paramfile.py", line 187, in __call__
    return self._check_for_uri_param(cli_argument, value)
  File "/usr/local/lib/python3.9/site-packages/awscli/paramfile.py", line 195, in _check_for_uri_param
    raise ParamError(param.cli_name, six.text_type(e))
awscli.argprocess.ParamError: Error parsing parameter '--invocation-endpoint': Unable to retrieve http://localhost:3300/test: Could not connect to the endpoint URL: "http://localhost:3300/test"
2022-12-20 19:26:19,711 - MainThread - awscli.clidriver - DEBUG - Exiting with rc 255
```

I also checked LocalStack logs, and the request was not even hitting LocalStack, which shows the issue came from the AWS CLI. (You can go look at the code for `paramfile.py` in the CLI code, it automatically checks if the hyperlink is fetchable). Which is why you would get the `hello` instead of the right parameter.

After passing the parameters enclosed in double + single quotes, here's the response:

```bash
{
    "ApiDestinationArn": "arn:aws:events:us-east-1:000000000000:api-destination/destination_1/8a115229-f77c-461d-a2f2-8f7ad8ba81f0",
    "ApiDestinationState": "ACTIVE",
    "CreationTime": 1671560934.938786,
    "LastModifiedTime": 1671560934.938786
}

```
And LocalStack logs, all good:
```bash
2022-12-20T19:28:54.941  INFO --- [   asgi_gw_0] localstack.request.aws     : AWS events.CreateApiDestination => 200; CreateApiDestinationRequest({'Name': 'destination_1', 'Description': 'send test event to local server', 'ConnectionArn': 'arn:aws:events:us-east-1:000000000000:connection/API_Connection_1/36285233-0b81-45c8-a28f-45fa6e43a395', 'InvocationEndpoint': "'http://localhost:3300/test'", 'HttpMethod': 'PUT', 'InvocationRateLimitPerSecond': 50}, headers={'Host': 'localhost:4566', 'Accept-Encoding': 'identity', 'X-Amz-Target': 'AWSEvents.CreateApiDestination', 'Content-Type': 'application/x-amz-json-1.1', 'User-Agent': 'aws-cli/1.24.5 Python/3.9.13 Darwin/22.1.0 botocore/1.26.5', 'X-Amz-Date': '20221220T182854Z', 'Authorization': 'AWS4-HMAC-SHA256 Credential=test/20221220/us-east-1/events/aws4_request, SignedHeaders=content-type;host;x-amz-date;x-amz-target, Signature=120c4177deedec2b819adb78ef61c033cd062199751508f3c7f7124b33715cd7', 'amz-sdk-invocation-id': 'c0f613c8-6507-4b0d-9ae9-c81232560710', 'amz-sdk-request': 'attempt=1', 'Content-Length': '309', 'x-localstack-tgt-api': 'events', 'x-moto-account-id': '000000000000'}); CreateApiDestinationResponse({'ApiDestinationArn': 'arn:aws:events:us-east-1:000000000000:api-destination/destination_1/8a115229-f77c-461d-a2f2-8f7ad8ba81f0', 'ApiDestinationState': 'ACTIVE', 'CreationTime': datetime.datetime(2022, 12, 20, 19, 28, 54, 938786, tzinfo=tzlocal()), 'LastModifiedTime': datetime.datetime(2022, 12, 20, 19, 28, 54, 938786, tzinfo=tzlocal())}, headers={'Content-Type': 'application/x-amz-json-1.1', 'Content-Length': '236', 'x-amzn-requestid': 'WIV5FDRXOCH833NKS0FDZWNZHZSZPRBDF0XEKCWTA2NTVS4H9G9X', 'Connection': 'close', 'Access-Control-Allow-Origin': '*', 'Access-Control-Allow-Methods': 'HEAD,GET,PUT,POST,DELETE,OPTIONS,PATCH', 'Access-Control-Allow-Headers': 'authorization,cache-control,content-length,content-md5,content-type,etag,location,x-amz-acl,x-amz-content-sha256,x-amz-date,x-amz-request-id,x-amz-security-token,x-amz-tagging,x-amz-target,x-amz-user-agent,x-amz-version-id,x-amzn-requestid,x-localstack-target,amz-sdk-invocation-id,amz-sdk-request', 'Access-Control-Expose-Headers': 'etag,x-amz-version-id'})
```

The endpoint is now the right one!

Hope that fixes your issue, please tell us if this works now.

**Comment by corinz** [comment]
thanks @bentsku !

**Comment by gage-russell** [comment]
Thank you @bentsku !! -- just one last hurdle in getting event bridge working with LocalStack over the line:

identifying a `NoneType has no attribute group` error took longer than it should have and was completely user-error, but worth putting somewhere on the internet. `connection-arn` is required to have a trailing slash based on this regex search in the LocalStack source code here: https://github.com/localstack/localstack/blob/09ce620b9b15580490a12f145dc9de8a55dc6c37/localstack/utils/aws/message_forwarding.py#L223

**Comment by bentsku** [comment]
Hello @gage-russell,

I think we should open a new issue to discuss this, is that possible? I would like to fix that and get a proper error message if that is reproducible. Thanks!

**Comment by Vuta** [comment]
thanks @bentsku, that indeed is the `awscli` issue. I'm closing this issue now. For anyone having the same problem, please refer to https://github.com/localstack/localstack/issues/6156#issuecomment-1359979937 for the fix.
Also, another thing worth mentioning here is that if you run localstack in a docker container, please register your local API destination's host as `host.docker.internal`. Otherwise local EventBridge won't be able to put event to your local server from inside docker. credit to @hoangvu03081
